test/gtest/ucp/test_ucp_atomic.cc

```diff
@@ -9,38 +9,69 @@
 
 extern "C" {
 #include <ucp/core/ucp_types.h> /* for atomic mode */
+#include <ucp/core/ucp_mm.h>
 }
 
 template <typename T>
 class test_ucp_atomic : public test_ucp_memheap {
 public:
-    static void get_test_variants(std::vector<ucp_test_variant>& variants)
+    /* Test variants */
+    enum {
+        ATOMIC_MODE  = UCS_MASK(2),
+        ENABLE_PROTO = UCS_BIT(2)
+    };
+
+    static void get_test_variants(std::vector<ucp_test_variant>& variants,
+                                  int variant, const std::string& name)
     {
         uint64_t features = (sizeof(T) == sizeof(uint32_t)) ?
                             UCP_FEATURE_AMO32 : UCP_FEATURE_AMO64;
-        add_variant_with_value(variants, features, UCP_ATOMIC_MODE_CPU, "cpu");
-        add_variant_with_value(variants, features, UCP_ATOMIC_MODE_DEVICE, "device");
-        add_variant_with_value(variants, features, UCP_ATOMIC_MODE_GUESS, "guess");
+        add_variant_with_value(variants, features, variant | UCP_ATOMIC_MODE_CPU, "cpu" + name);
+        add_variant_with_value(variants, features, variant | UCP_ATOMIC_MODE_DEVICE, "device" + name);
+        add_variant_with_value(variants, features, variant | UCP_ATOMIC_MODE_GUESS, "guess" + name);
+    }
+
+    static void get_test_variants(std::vector<ucp_test_variant>& variants)
+    {
+        get_test_variants(variants, 0, "");
+        get_test_variants(variants, ENABLE_PROTO, "/proto");
     }
 
+    struct send_func_data {
+        ucp_atomic_op_t op;
+        ucs_memory_type_t send_mem_type;
+        ucs_memory_type_t recv_mem_type;
+    };
+
     void post(size_t size, void *target_ptr, ucp_rkey_h rkey,
               void *expected_data, void *arg)
     {
-        ucp_atomic_op_t op = *(ucp_atomic_op_t*)arg;
-        T value            = (T)ucs::rand() * (T)ucs::rand();
-        T prev             = *(T*)target_ptr;
-        *(T*)expected_data = atomic_op_result(op, value, prev, 0);
+        send_func_data* data = (send_func_data*)arg;
+        T value              = (T)ucs::rand() * (T)ucs::rand();
+        T prev;
+        T result;
+
+        mem_buffer::copy_from(&prev, target_ptr, sizeof(T),
+                              data->recv_mem_type);
+        result = atomic_op_result(data->op, value, prev, 0);
+        mem_buffer::copy_to(expected_data, &result, sizeof(T),
+                            data->send_mem_type);
 
         ucp_request_param_t param;
         param.op_attr_mask = 0;
-        ucs_status_t status = do_atomic(op, size, target_ptr, rkey, value, param);
+        ucs_status_t status = do_atomic(data->op, size, target_ptr, rkey,
```

Comment:
do_atomic does not really use mem type, it always uses cpu memory
need to pass pointer to it, which is mem_buffer